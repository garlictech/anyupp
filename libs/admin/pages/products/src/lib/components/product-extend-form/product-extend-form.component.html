<nb-card>
  <nb-card-header>
    <div class="float-start">
      {{
        (editing ? 'products.editProduct' : 'products.extendProduct')
          | translate
      }}
    </div>
    <button
      nbButton
      ghost
      shape="round"
      size="small"
      class="float-right"
      (click)="close()"
    >
      <nb-icon icon="close-outline"></nb-icon>
    </button>
  </nb-card-header>
  <nb-card-body>
    <form [formGroup]="dialogForm" (ngSubmit)="submit()" autocomplete="off">
      <fieldset class="form">
        <legend class="label">{{ 'products.basicInfo' | translate }}</legend>

        <div class="subtitle-2 mb-2">
          <span>{{ product?.name | localize }}</span>
        </div>
        <div class="caption mb-2">
          <span>{{ product?.description | localize }}</span>
        </div>

        <div class="row">
          <div class="col-3" *ngIf="productLevel === eProductLevel.GROUP">
            <bgap-form-text-input
              caption="common.taxPercent"
              type="number"
              [control]="dialogForm.get('tax')! | asFormControl"
              inputId="tax"
            ></bgap-form-text-input>
          </div>
          <div class="col-3" *ngIf="productLevel === eProductLevel.GROUP">
            <bgap-form-text-input
              caption="common.takeawayTaxPercent"
              type="number"
              [control]="dialogForm.get('takeawayTax')! | asFormControl"
              inputId="takeawayTax"
            ></bgap-form-text-input>
          </div>
          <div class="col-3" *ngIf="productLevel === eProductLevel.UNIT">
            <bgap-form-select
              [control]="dialogForm.get('laneId')! | asFormControl"
              caption="products.lane"
              [options]="(this.unitLanes$ | async) || []"
              [localize]="false"
              [forceShowEmptyOption]="true"
              inputId="laneId"
            >
            </bgap-form-select>
          </div>

          <div class="col-3 pt-4">
            <bgap-form-checkbox
              caption="products.isVisible"
              [control]="dialogForm.get('isVisible')! | asFormControl"
              inputId="isVisible"
            ></bgap-form-checkbox>
          </div>
        </div>
      </fieldset>

      <bgap-form-serving-mode
        *ngIf="productLevel === eProductLevel.UNIT"
        [control]="dialogForm.get('supportedServingModes')! | asFormControl"
        [servingModes]="servingModes"
      ></bgap-form-serving-mode>

      <bgap-form-product-variants
        [productLevel]="productLevel"
        [variantFormArray]="dialogForm.get('variants')! | asFormArray"
        [allowAddVariant]="false"
        [currency]="currency"
      ></bgap-form-product-variants>

      <bgap-form-product-components
        [productLevel]="productLevel"
        [componentFormArray]="dialogForm.get('configSets')! | asFormArray"
        [currency]="currency"
      ></bgap-form-product-components>

      <div class="mt-20">
        <button
          nbButton
          size="small"
          type="submit"
          fullWidth
          status="primary"
          [disabled]="dialogForm.invalid"
        >
          {{ 'common.submit' | translate }}
        </button>
      </div>
    </form>
  </nb-card-body>
</nb-card>
