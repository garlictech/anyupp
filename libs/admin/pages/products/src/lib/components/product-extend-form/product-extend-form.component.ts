import * as fp from 'lodash/fp';
import { NGXLogger } from 'ngx-logger';
import { Observable } from 'rxjs';
import { take } from 'rxjs/operators';

import {
  ChangeDetectionStrategy,
  Component,
  Injector,
  OnInit,
} from '@angular/core';
import { FormArray } from '@angular/forms';
import { loggedUserSelectors } from '@bgap/admin/shared/data-access/logged-user';
import { productCategoriesSelectors } from '@bgap/admin/shared/data-access/product-categories';
import { unitsSelectors } from '@bgap/admin/shared/data-access/units';
import { AbstractFormDialogComponent } from '@bgap/admin/shared/forms';
import { EToasterType } from '@bgap/admin/shared/utils';
import { EProductLevel, IKeyValue, Product } from '@bgap/shared/types';
import { cleanObject, filterNullish } from '@bgap/shared/utils';
import { UntilDestroy, untilDestroyed } from '@ngneat/until-destroy';
import { select, Store } from '@ngrx/store';
import { CrudSdkService } from '@bgap/admin/shared/data-access/data';
import * as CrudApi from '@bgap/crud-gql/api';

import { ProductFormService } from '../../services/product-form/product-form.service';

@UntilDestroy()
@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  selector: 'bgap-product-extend-form',
  templateUrl: './product-extend-form.component.html',
})
export class ProductExtendFormComponent
  extends AbstractFormDialogComponent
  implements OnInit {
  public product?: Product;
  public productLevel!: EProductLevel;
  public eProductLevel = EProductLevel;
  public editing = false;
  public currency!: string;
  public productCategories$: Observable<CrudApi.ProductCategory[]>;
  public unitLanes: IKeyValue[] = [];

  private _selectedChainId?: string;
  private _selectedGroupId?: string;
  private _selectedUnitId?: string;

  constructor(
    protected _injector: Injector,
    private crudSdk: CrudSdkService,
    private _store: Store,
    private _productFormService: ProductFormService,
    private _logger: NGXLogger,
  ) {
    super(_injector);

    this._store
      .pipe(
        select(loggedUserSelectors.getLoggedUserSettings),
        take(1),
        filterNullish(),
      )
      .subscribe((userSettings: CrudApi.AdminUserSettings): void => {
        this._selectedChainId = userSettings?.selectedChainId || '';
        this._selectedGroupId = userSettings?.selectedGroupId || '';
        this._selectedUnitId = userSettings?.selectedUnitId || '';
      });

    this.productCategories$ = this._store.pipe(
      select(productCategoriesSelectors.getAllProductCategories),
      untilDestroyed(this),
    );

    this._store
      .pipe(select(unitsSelectors.getSelectedUnit), filterNullish(), take(1))
      .subscribe(
        unit =>
          (this.unitLanes = unit.lanes
            ? unit.lanes.map(lane => ({
                key: lane?.id || '',
                value: lane?.name,
              }))
            : []),
      );
  }

  ngOnInit(): void {
    this.dialogForm = this._productFormService.createProductExtendFormGroup(
      this.productLevel,
    );

    if (this.product) {
      this.dialogForm.patchValue(
        fp.omit(['variants', 'configSets'], cleanObject(this.product)),
      );

      this._productFormService.patchExtendedProductVariants(
        this.product,
        this.dialogForm?.controls.variants as FormArray,
      );

      this._productFormService.patchConfigSet(
        this.product,
        this.productLevel,
        this.dialogForm?.controls.configSets as FormArray,
      );
    } else {
      this.dialogForm.controls.isVisible.patchValue(true);

      if (this.productLevel === EProductLevel.UNIT) {
        this.dialogForm.controls.takeaway.patchValue(false);
      }
    }
  }

  public async submit(): Promise<void> {
    if (this.dialogForm?.valid) {
      const value = { ...this.dialogForm?.value };

      if (!this.product) {
        throw new Error('HANDLE ME: product cannot be undefined');
      }

      if (this.editing) {
        try {
          const input = !{
            input: {
              id: this.product.id,
              ...value,
            },
          };

          if (!input) {
            throw new Error('HANDLE ME: input cannot be undefined');
          }

          if (this.productLevel === EProductLevel.GROUP) {
            await this.crudSdk.sdk.UpdateGroupProduct(input).toPromise();
          } else {
            await this.crudSdk.sdk.UpdateUnitProduct(input).toPromise();
          }

          this._toasterService.show(
            EToasterType.SUCCESS,
            '',
            'common.updateSuccessful',
          );
          this.close();
        } catch (error) {
          this._logger.error(
            `EXTENDED PRODUCT UPDATE ERROR: ${JSON.stringify(error)}`,
          );
        }
      } else {
        // Save the extended product id
        value.parentId = this.product.id;
        value.chainId = this._selectedChainId;
        value.groupId = this._selectedGroupId;
        if (this.productLevel === EProductLevel.UNIT) {
          value.unitId = this._selectedUnitId;
          value.position = 0;
        }

        try {
          const input = {
            input: value,
          };

          if (this.productLevel === EProductLevel.GROUP) {
            await this.crudSdk.sdk.CreateGroupProduct(input).toPromise();
          } else {
            await this.crudSdk.sdk.CreateUnitProduct(input).toPromise();
          }

          this._toasterService.show(
            EToasterType.SUCCESS,
            '',
            'common.insertSuccessful',
          );
          this.close();
        } catch (error) {
          this._logger.error(
            `EXTENDED PRODUCT INSERT ERROR: ${JSON.stringify(error)}`,
          );
        }
      }
    }
  }
}
