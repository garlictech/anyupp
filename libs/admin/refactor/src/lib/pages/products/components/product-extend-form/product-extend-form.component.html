<nb-card>
  <nb-card-header>
    <div class="float-start">
      {{ 'products.editProduct' | translate }}
    </div>
    <button
      nbButton
      ghost
      shape="round"
      size="small"
      class="float-right"
      (click)="close()"
    >
      <nb-icon icon="close" pack="material-icons"></nb-icon>
    </button>
  </nb-card-header>
  <nb-card-body>
    <form
      *ngIf="dialogForm"
      [formGroup]="dialogForm"
      (ngSubmit)="submit()"
      autocomplete="off"
    >
      <fieldset class="form">
        <legend class="label">{{ 'products.basicInfo' | translate }}</legend>

        <div class="subtitle-2 mb-2">
          <span>{{ product?.name | localize }}</span>
        </div>
        <div class="caption mb-2">
          <span>{{ product?.description | localize }}</span>
        </div>

        <div class="row">
          <div class="col-3">
            <bgap-form-text-input
              caption="common.taxPercent"
              type="number"
              [control]="dialogForm?.get('tax') | asFormControl"
              inputId="tax"
              suffix="%"
            ></bgap-form-text-input>
          </div>
          <div class="col-3">
            <bgap-form-text-input
              caption="common.takeawayTaxPercent"
              type="number"
              [control]="dialogForm?.get('takeawayTax') | asFormControl"
              inputId="takeawayTax"
              suffix="%"
            ></bgap-form-text-input>
          </div>
          <div class="col-3">
            <bgap-form-select
              [control]="dialogForm?.get('laneId') | asFormControl"
              caption="products.lane"
              [options]="(this.unitLanes$ | async) || []"
              [localize]="false"
              [forceShowEmptyOption]="true"
              inputId="laneId"
            >
            </bgap-form-select>
          </div>

          <div class="col-3 pt-4">
            <bgap-form-checkbox
              caption="products.isVisible"
              [control]="dialogForm?.get('isVisible') | asFormControl"
              inputId="isVisible"
            ></bgap-form-checkbox>
          </div>
        </div>
      </fieldset>

      <bgap-form-serving-mode
        [control]="dialogForm?.get('supportedServingModes') | asFormControl"
        [servingModes]="servingModes"
      ></bgap-form-serving-mode>

      <bgap-form-product-variants
        [variantFormArray]="dialogForm?.get('variants') | asFormArray"
        [allowAddVariant]="false"
        [currency]="currency"
        [unitServiceFeePolicy]="selectedUnit?.serviceFeePolicy"
        [productTax]="product?.tax"
      ></bgap-form-product-variants>

      <bgap-form-product-components
        [componentFormArray]="dialogForm?.get('configSets') | asFormArray"
        [currency]="currency"
      ></bgap-form-product-components>

      <div class="mt-20">
        <button
          nbButton
          size="small"
          type="submit"
          fullWidth
          status="primary"
          [disabled]="dialogForm?.invalid || (working$ | async)"
        >
          {{ 'common.submit' | translate }}
        </button>
      </div>
    </form>
  </nb-card-body>
</nb-card>
