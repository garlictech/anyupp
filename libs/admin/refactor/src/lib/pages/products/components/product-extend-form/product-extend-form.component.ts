import * as fp from 'lodash/fp';
import { iif, Observable } from 'rxjs';
import { switchMap, take, tap } from 'rxjs/operators';

import {
  ChangeDetectionStrategy,
  Component,
  Injector,
  OnInit,
} from '@angular/core';
import { FormArray } from '@angular/forms';
import { AdminUserSettings, ProductCategory, ServingMode } from '@bgap/domain';
import {
  EProductLevel,
  KeyValue,
  Product,
  UpsertResponse,
} from '@bgap/shared/types';
import { cleanObject, filterNullish } from '@bgap/shared/utils';
import { UntilDestroy } from '@ngneat/until-destroy';
import { select } from '@ngrx/store';

import { AbstractFormDialogComponent } from '../../../../shared/forms';
import { SERVING_MODES } from '../../../../shared/utils';
import { loggedUserSelectors } from '../../../../store/logged-user';
import { ProductCategoryCollectionService } from '../../../../store/product-categories';
import { unitsSelectors } from '../../../../store/units';
import { ProductFormService } from '../../services/product-form.service';

@UntilDestroy()
@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  selector: 'bgap-product-extend-form',
  templateUrl: './product-extend-form.component.html',
})
export class ProductExtendFormComponent
  extends AbstractFormDialogComponent
  implements OnInit
{
  public product?: Product;
  public productLevel!: EProductLevel;
  public eProductLevel = EProductLevel;
  public editing = false;
  public currency!: string;
  public productCategories$: Observable<ProductCategory[]>;
  public unitLanes$: Observable<KeyValue[]>;
  public servingModes = SERVING_MODES;
  public selectedUnit?: CrudApi.Unit;

  private _selectedChainId = '';
  private _selectedGroupId = '';
  private _selectedUnitId = '';

  constructor(
    protected override _injector: Injector,
    private _productFormService: ProductFormService,
    private _productCategoryCollectionService: ProductCategoryCollectionService,
  ) {
    super(_injector);

    this._store
      .pipe(
        select(loggedUserSelectors.getLoggedUserSettings),
        take(1),
        filterNullish(),
      )
      .subscribe((userSettings: AdminUserSettings) => {
        this._selectedChainId = userSettings?.selectedChainId || '';
        this._selectedGroupId = userSettings?.selectedGroupId || '';
        this._selectedUnitId = userSettings?.selectedUnitId || '';
      });

    this.productCategories$ =
      this._productCategoryCollectionService.filteredEntities$;

    this.unitLanes$ = this._store.pipe(
      select(unitsSelectors.getSelectedUnitLanes),
      filterNullish(),
      take(1),
    );

    this._store
      .pipe(select(unitsSelectors.getSelectedUnit), filterNullish(), take(1))
      .subscribe(unit => {
        this.selectedUnit = unit;
      });
  }

  ngOnInit() {
    this.dialogForm = this._productFormService.createProductExtendFormGroup(
      this.productLevel,
    );

    if (this.product) {
      this.dialogForm.patchValue(
        fp.omit(['variants', 'configSets'], cleanObject(this.product)),
      );

      this._productFormService.patchExtendedProductVariants(
        this.product.variants || [],
        this.dialogForm?.controls['variants'] as FormArray,
      );

      this._productFormService.patchConfigSet(
        this.product.configSets || [],
        this.dialogForm?.controls['configSets'] as FormArray,
      );
    } else {
      this.dialogForm.controls['isVisible'].patchValue(true);
    }
  }

  public submit() {
    if (this.dialogForm?.valid) {
      const takeaway = (
        <string[]>this.dialogForm.value.supportedServingModes || []
      ).includes(ServingMode.takeaway);

      const parentInfo = {
        parentId: this.product?.id || '',
        chainId: this._selectedChainId,
        groupId: this._selectedGroupId,
      };

      this.setWorking$(true)
        .pipe(
          switchMap(() =>
            iif(
              () => this.productLevel === EProductLevel.UNIT,
              this._productFormService.saveUnitExtendForm$(
                {
                  ...this.dialogForm?.value,
                  ...parentInfo,
                  unitId: this._selectedUnitId,
                  position: 0,
                  takeaway,
                },
                {
                  ...this.dialogForm?.value,
                  id: this.product?.id || '',
                  dirty: this.product?.dirty ? false : undefined,
                  takeaway,
                },
                this.editing,
              ),
              this._productFormService.saveGroupExtendForm$(
                {
                  ...this.dialogForm?.value,
                  ...parentInfo,
                  takeawayTax: this.dialogForm?.value.takeawayTax || null, // save or remove
                },
                {
                  ...this.dialogForm?.value,
                  id: this.product?.id || '',
                  dirty: this.product?.dirty ? false : undefined, // save or leave
                  takeawayTax: this.dialogForm?.value.takeawayTax || null, // save or remove
                },
                this.editing,
              ),
            ),
          ),
          tap(() => this.setWorking$(false)),
        )
        .subscribe((response: UpsertResponse<unknown>) => {
          this._toasterService.showSimpleSuccess(response.type);

          this.close();
        });
    }
  }
}
