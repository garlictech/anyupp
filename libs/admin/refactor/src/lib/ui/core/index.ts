export * from './admin-ui-core.module';
export * from './utils';
