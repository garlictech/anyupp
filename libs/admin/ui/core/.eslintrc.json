{
  "extends": "../../../../.eslintrc.json",
  "ignorePatterns": ["!**/*"],
  "rules": {},
  "overrides": [
    {
      "files": [
        "*.ts"
      ],
      "plugins": [
        "@typescript-eslint"
      ],
      "parserOptions": {
        "project": [
          "libs/admin/ui/core/tsconfig.*?.json"
        ]
      }
    }
  ]
}
