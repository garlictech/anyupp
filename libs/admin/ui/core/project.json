{
  "projectType": "library",
  "root": "libs/admin/ui/core",
  "sourceRoot": "libs/admin/ui/core/src",
  "prefix": "bgap",
  "targets": {
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "options": {
        "lintFilePatterns": ["libs/admin/ui/core/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["coverage/libs/admin/ui/core"],
      "options": {
        "jestConfig": "libs/admin/ui/core/jest.config.js",
        "passWithNoTests": true
      }
    }
  },
  "tags": ["platform:angular"]
}
