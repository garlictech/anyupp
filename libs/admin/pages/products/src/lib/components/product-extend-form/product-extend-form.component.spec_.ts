/*
import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';

import { ProductExtendFormComponent } from './product-extend-form.component';

xdescribe('ProductExtendFormComponent', (): void => {
  let component: ProductExtendFormComponent;
  let fixture: ComponentFixture<ProductExtendFormComponent>;

  beforeEach(
    waitForAsync((): void => {
      TestBed.configureTestingModule({
        declarations: [ProductExtendFormComponent],
      }).compileComponents();
    }),
  );

  beforeEach((): void => {
    fixture = TestBed.createComponent(ProductExtendFormComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', (): void => {
    expect(component).toBeTruthy();
  });
});
*/
