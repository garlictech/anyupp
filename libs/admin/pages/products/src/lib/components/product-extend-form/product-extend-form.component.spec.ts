import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';

import { ProductExtendFormComponent } from './product-extend-form.component';

describe('ProductExtendFormComponent', (): void => {
  let component: ProductExtendFormComponent;
  let fixture: ComponentFixture<ProductExtendFormComponent>;

  beforeEach(
    waitForAsync((): void => {
      TestBed.configureTestingModule({
        declarations: [ProductExtendFormComponent],
      }).compileComponents();
    })
  );

  beforeEach((): void => {
    fixture = TestBed.createComponent(ProductExtendFormComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', (): void => {
    expect(component).toBeTruthy();
  });
});
