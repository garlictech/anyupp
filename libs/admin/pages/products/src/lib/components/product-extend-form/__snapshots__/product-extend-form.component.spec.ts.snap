// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductExtendFormComponent Group Form should create empty group product form 1`] = `
Object {
  "configSets": Array [],
  "isVisible": true,
  "takeawayTax": "",
  "tax": "",
  "variants": Array [],
}
`;

exports[`ProductExtendFormComponent Unit Form should create empty unit product form 1`] = `
Object {
  "configSets": Array [],
  "isVisible": true,
  "laneId": "",
  "supportedServingModes": Array [],
  "variants": Array [],
}
`;
